### PR TITLE
Add option to set timeout (resolves #22)

### LINKED ISSUE #22: Timeout is not configurable
It might be useful for some users to set the network timeout to a custom value. (Someone requested this for Neo4j::Driver.) This isn’t currently possible with Neo4j::Bolt.

It looks like adding this would require calling [`neo4j_new_config()`](https://neo4j-client.net/doc/latest/neo4j-client_8h.html#ade0b60861904e281bbe8076a73ab1b60), which could then be passed to [`neo4j_connect()`](https://neo4j-client.net/doc/latest/neo4j-client_8h.html#a84dff5eae1df5b5b27cc160ca867d467). However, unlike most other config options, `connect_timeout` doesn’t appear to have library functions that modify it. We may need to adjust the [`neo4j_config struct`](https://github.com/cleishm/libneo4j-client/blob/master/lib/src/client_config.h#L23) manually.

### PR DESCRIPTION
**Pull request by johannessen**
I took a swing at the timeout.

The problem is, the `neo4j_config` struct’s definition differs based on `#ifdef` directives (see [client_config.h](https://github.com/cleishm/libneo4j-client/blob/master/lib/src/client_config.h#L23)). We can’t copy it verbatim. The “solution” I came up with is to only use half the struct’s definition, so the size is off. For this particular problem, that doesn’t really matter because the timeout’s position in the struct is before the point where the struct may start to differ.

But of course, as soon as you `sizeof()` this struct in any way, it blows up in your face.

I don’t like this much. But I don’t know another way to solve it. Do you?

### CODE REVIEW

**Review by majensen — COMMENTED** [review]
I wouldn't shadow the config struct; it gets appropriately allocated and loaded with defaults in neo4j_new_config(). The only variation w/r configure I see is whether TLS is activated or not - in which case, the tls fields are not present in the structure. This issue is handled in the setters (neo4j_config_set_TLS_*); they error gracefully when TLS is not compiled in. 

However, time_t connect_timeout is always in the config. It doesn't have a setter, but simple assignment (with a (time_t) cast?) should do the trick. 

TL;DR - zap l19-l38 and I think we're good.

**Comment by johannessen** [comment]
> zap l19-l38 and I think we're good.

I did try that, but sadly it results in:

    Bolt.xs:60:9: error: incomplete definition of type 'struct neo4j_config'
      config->connect_timeout = (time_t)timeout;
      ~~~~~~^
    .../neo4j-client.h:63:16: note: 
          forward declaration of 'struct neo4j_config'

I see now that on my system, it’s apparently enough to just `#include <client_config.h>`. The struct is defined in that header file. But is this portable to other environments? Is `client_config.h` indeed always available when libneo4j-client is installed?

I’m a bit out of my depth here, as I have very little experience with C. Your help is greatly appreciated.

Good point about making the cast explicit, BTW.

**Comment by johannessen** [comment]
[Travis failed](https://travis-ci.org/majensen/perlbolt/jobs/652687182) with `client_config.h: No such file`, so evidently this is indeed *not* portable.

Back to plan A then, I suppose…? 😕

**Comment by majensen** [comment]
I think the real problem may be, we may be one of the few projects that actually put libneo4j-client through its paces. What it needs is a getter/setter pair like neo4j_config_get_timeout/neo4j_config_set_timeout. This would operate in the client_config.c scope, and would just work like neo4j_config_TLS_* - no knowledge of struct config required. I think it's probably just an oversight. I note that 'timeout' is a field that doesn't have a default set in new_neo4j_config(). 

So (1) a helpful thing would be creating the getter/setter in a libneo4j-client fork and submitting a pull request. That doesn't solve our problem immediately, of course. 
(2) Do the practical thing and copy the declaration. Rather than burying it in the module code, I think it may be easier to remember later that we did this if we create a special header file in our code, and #include that in the module code. Even though we don't know the config, we can test whether TLS was in the compile by calling one of the TLS getters, which fail gracefully. That would tell us how to set HAVE_TLS. This info has to get to the build -- hmm.
Let me play with it for a bit.

**Comment by majensen** [comment]
@johannessen thoughts? This solution doesn't check for the TLS, so it's still funky. Have to think more about how to do that.

**Review by johannessen — COMMENTED** [review]
Looks good to me so far. The special header file is a nice idea!

**Comment by majensen** [comment]
All right! I'll merge this in and we can put the "HAVE_TLS" issue in the backlog...
